pre_commit/languages/pygrep.py

```diff
@@ -75,6 +103,15 @@ def main(argv: Optional[Sequence[str]] = None) -> int:
     pattern = re.compile(args.pattern.encode(), flags)
 
     retv = 0
+
+    if args.negate:
+        for filename in args.filenames:
+            if args.multiline:
+                retv |= _process_filename_at_once_negated(pattern, filename)
+            else:
+                retv |= _process_filename_by_line_negated(pattern, filename)
+        return retv
```

Comment:
you can probably do:

```
# probably at module scope
class Choice(NamedTuple):
    multiline: bool
    negate: bool
FNS = {
    Choice(multiline=True, negate=True): _process_filename_at_once_negated,
    Choice(multiline=True, negate=False): _process_filename_at_once,
    Choice(multiline=False, negate=True): _process_filename_by_line_negated,
    Choice(multiline=False, negate=False): _process_filename_by_line,
}
# in code
process_fn = FNS[Choice(multiline=args.multiline, negate=args.negate)]

# then the for loop code, but without branches
```

just because there's a lot more duplication than before

Reply:
@asottile yes, that's much cleaner, thanks!